pkg/kubelet/rkt/rkt.go

```diff
@@ -1315,45 +1347,124 @@ func (r *Runtime) SyncPod(pod *api.Pod, podStatus api.PodStatus, internalPodStat
-	if err := exec.Command("systemctl", "reset-failed").Run(); err != nil {
-		glog.Errorf("rkt: Failed to reset failed systemd services: %v, continue to gc anyway...", err)
-	}
+	glog.V(4).Infof("rkt: Garbage collecting triggered with policy %v", gcPolicy)
 
-	if _, err := r.runCommand("gc", "--grace-period="+gcPolicy.MinAge.String(), "--expire-prepared="+gcPolicy.MinAge.String()); err != nil {
-		glog.Errorf("rkt: Failed to gc: %v", err)
+	if err := r.systemd.ResetFailed(); err != nil {
+		glog.Errorf("rkt: Failed to reset failed systemd services: %v, continue to gc anyway...", err)
 	}
 
-	// GC all inactive systemd service files.
-	units, err := r.systemd.ListUnits()
+	resp, err := r.apisvc.ListPods(context.Background(), &rktapi.ListPodsRequest{
+		Detail:  false,
+		Filters: kubernetesPodsFilters(),
+	})
 	if err != nil {
-		glog.Errorf("rkt: Failed to list units: %v", err)
-		return err
+		return fmt.Errorf("rkt: Failed to get dead pods list: %v", err)
 	}
-	runningKubernetesUnits := sets.NewString()
-	for _, u := range units {
-		if strings.HasPrefix(u.Name, kubernetesUnitPrefix) && u.SubState == "running" {
-			runningKubernetesUnits.Insert(u.Name)
+
+	var totalContainers int
+	var inactivePods []*rktapi.Pod
+	var removeCandidates []*rktapi.Pod
+
+	// Deleted pods will be GC'd anyway.
+	for _, pod := range resp.Pods {
+		if pod.State != rktapi.PodState_POD_STATE_RUNNING {
+			uid := getPodUID(pod)
+			if uid == types.UID("") {
+				glog.Errorf("rkt: Cannot get the UID of pod %q, will remove it", pod.Id)
+				removeCandidates = append(removeCandidates, pod)
+				continue
+			}
+			_, found := r.podGetter.GetPodByUID(uid)
+			if !found {
+				removeCandidates = append(removeCandidates, pod)
+				continue
+			}
+
+			inactivePods = append(inactivePods, pod)
+			totalContainers = totalContainers + len(pod.Apps)
 		}
 	}
 
-	files, err := ioutil.ReadDir(systemdServiceDir)
-	if err != nil {
-		glog.Errorf("rkt: Failed to read the systemd service directory: %v", err)
-		return err
+	sort.Sort(podsByCreatedAt(inactivePods))
+
+	// Enforce GCPolicy.MaxContainers.
+	for _, pod := range inactivePods {
+		if totalContainers <= gcPolicy.MaxContainers {
+			break
+		}
+		creationTime := time.Unix(0, pod.CreatedAt)
+		if creationTime.Add(gcPolicy.MinAge).Before(time.Now()) {
+			// The pod is old and we are exceeding the MaxContainers limit.
+			// Delete the pod.
+			removeCandidates = append(removeCandidates, pod)
+			totalContainers = totalContainers - len(pod.Apps)
+		}
 	}
-	for _, f := range files {
-		if strings.HasPrefix(f.Name(), kubernetesUnitPrefix) && !runningKubernetesUnits.Has(f.Name()) && f.ModTime().Before(time.Now().Add(-gcPolicy.MinAge)) {
-			glog.V(4).Infof("rkt: Removing inactive systemd service file: %v", f.Name())
-			if err := os.Remove(serviceFilePath(f.Name())); err != nil {
-				glog.Warningf("rkt: Failed to remove inactive systemd service file %v: %v", f.Name(), err)
+
+	// Remove in parallel.
+	var wg sync.WaitGroup
```

Comment:
Is there a reason why we should remove in parallel here? Presumably this operation should be mostly fs bound, meaning that in parallel shouldn't be that much faster.. and it's not in the k8 critical path regardless.

Alternately, to remove in parallel (at the expense of good error messages) we could exec `rkt rm` with multiple uuids and batch removes that way (though I don't think that's worth doing).

Happy to hear a reason why this should be in parallel / any experience here, but I'd personally vote for just sequential removal for simplicity.